### PR TITLE
Chore: improve autofix performance (fixes #9998)

### LINKED ISSUE #9998: Avoid running non-autofixable rules multiple times when doing multipass autofixing
In ESLint v4.18.1,  our autofixing logic looks roughly like this (pseudocode):

1. Parse the code.
1. Run all the enabled rules on the code, and get the linting errors.
1. If there are any autofixable errors*, apply the autofixes to get a new version of the source code, and start over from step 1 on the new code.
1. Return the linting errors and autofixed code from the last run.

<sup>*This loop also has a counter which breaks after a fixed number of iterations to ensure that it halts at some point.</sup>

This has a disadvantage that the enabled rules are frequently run more than once. With autofixable rules, the result of each run is always used to produce the source text for the next run. However, with non-autofixable rules, the result of running the rule is discarded until the final pass.

I think we could get a significant performance improvement if we only run autofixable rules on the first few passes, until no more fixes are possible. Afterwards, we would run all the non-autofixable rules on the final AST by doing a second traversal, and combine their messages with the messages from the fixable rules.

### PR DESCRIPTION
**Pull request by not-an-aardvark**
**What is the purpose of this pull request? (put an "X" next to item)**

[x] Other, please explain:

**What changes did you make? (Give an overview)**

This updates the multipass autofix logic to only run non-autofixable rules once on the last pass, rather than running them on every pass and discarding everything except the last results. ESLint now only runs autofixable rules on the first few passes, until no more fixes are possible. Afterwards, it runs all of the non-autofixable rules on the final AST by doing a second traversal, and combines their messages with the messages from the fixable rules. (See #9998 for a bit more information on how this affects the autofixing logic.)

This results in a performance improvement of about 20-25% when running autofix mode. I tested this by linting the ESLint codebase with a few rules modified (to ensure that autofixable errors are reported and fixed):

```
time eslint lib/ tests/ bin/ conf/ tools/ Makefile.js .eslintrc.js --rule 'curly: [error, multi]' --rule 'nonblock-statement-body-position: [error, beside]' --fix-dry-run
```

This increases the internal complexity of `linter`, because it needs to be able to run two sets of rules on the same AST at different times, without reparsing the source code or reprocessing all of the directive comments and environments.

**Is there anything you'd like reviewers to focus on?**

Is the performance improvement worth the additional complexity?

### CODE REVIEW

File: lib/linter.js [hunk 1]
```diff
@@ -736,10 +769,16 @@ const BASE_TRAVERSAL_CONTEXT = Object.freeze(
...
-function runRules(sourceCode, configuredRules, ruleMapper, parserOptions, parserName, settings, filename) {
+function runRules(traversalParameters) {
     const emitter = createEmitter();
     const traverser = new Traverser();
 
+    const sourceCode = traversalParameters.sourceCode;
+    const parserOptions = traversalParameters.parserOptions;
+    const parserName = traversalParameters.parserName;
+    const settings = traversalParameters.settings;
+    const filename = traversalParameters.filename;
```
Comment (not-an-aardvark) [inline lib/linter.js]: I eagerly await being able to can use destructuring in this codebase when we drop Node 4 support 😄

**Comment by ilyavolodin** [comment]
Could you explain what happens (and if the behavior is any different from before), if autofixing causes an error in another rule?

**Comment by not-an-aardvark** [comment]
> I thought before we only committed autofixed code if the number of errors didn't increase.

This is not correct. If autofixing causes a non-autofixable error in another rule, we simply report the error and commit the autofix (when there are no more fixes to apply). If autofixing causes a fixable error in another rule, that error is usually fixed on the next pass. This PR doesn't change the observable behavior around which errors get fixed.

**Comment by ilyavolodin** [comment]
Hmm.. Interesting. I'm pretty sure that in the very first implementation of autofixes, we never committed fixed code if it caused more errors, but I guess it changed somewhere along the way.

File: lib/linter.js [hunk 2]
```diff
@@ -727,7 +760,7 @@ const BASE_TRAVERSAL_CONTEXT = Object.freeze(
 
 /**
  * Runs the given rules on the given SourceCode object
- * @param {SourceCode} sourceCode A SourceCode object for the given text
+ * @param {TraversalParameters} traversalParameters The TraversalParameters for the given run
```
Comment (alberto) [inline lib/linter.js]: You should remove the following params from the doc.

**Comment by platinumazure** [comment]
@ilyavolodin How would you know if there were more errors? Just looking at the count would be one thing, but it would be very common for errors to move around as code elsewhere in the file is autofixed so I don't know how we could identify "new" vs "preexisting" errors.

**Comment by platinumazure** [comment]
Overall, I'm not a huge fan of the extra complexity here. The 20-25% extra performance gain is *tempting*, but I wonder if we could do more to encapsulate running all rules vs running fixable rules only.

I feel like this is something I would love to pursue if we could first do some more refactoring of Linter for readability.

That's just my take. If you want to keep pursuing the implementation, maybe it could be discussed in a TSC meeting?

**Comment by not-an-aardvark** [comment]
Thanks, that seems like a reasonable take. I am a bit skeptical that we would be able to significantly simplify this by refactoring `Linter` more -- I think there is a decent amount of complexity inherent to this approach. For example, this approach requires setting the `reportUnusedDisableDirectives` option to two separate values on the different passes over the AST to ensure that directives aren't considered "unused" when only running fixable rules (since there might still be warnings from non-fixable rules that cause the directive to be used). That said, it might be better to just not have this change at all due to the complexity.

I'll plan to close this in a few days if no one thinks it's worth changing.

**Pull request closed without merging**